libraries/hotstuff/include/eosio/hotstuff/qc_chain.hpp

```diff
@@ -0,0 +1,181 @@
+
+      bool am_i_proposer(); //check if I am the current proposer
+      bool am_i_leader(); //check if I am the current leader
+      bool am_i_finalizer(); //check if I am one of the current finalizers
+
+      void process_proposal(const hs_proposal_message& msg); //handles proposal
+      void process_vote(const hs_vote_message& msg); //handles vote
+      void process_new_view(const hs_new_view_message& msg); //handles new view
+      void process_new_block(const hs_new_block_message& msg); //handles new block
+
+      hs_vote_message sign_proposal(const hs_proposal_message& proposal, name finalizer); //sign proposal
+
+      bool extends(const fc::sha256& descendant, const fc::sha256& ancestor); //verify that a proposal descends from another
+
+      bool update_high_qc(const eosio::chain::quorum_certificate& high_qc); //check if update to our high qc is required
+
+      void leader_rotation_check(); //check if leader rotation is required
+
+      bool is_node_safe(const hs_proposal_message& proposal); //verify if a proposal should be signed
+
+      std::vector<hs_proposal_message> get_qc_chain(const fc::sha256& proposal_id); //get 3-phase proposal justification
+
+      void send_hs_proposal_msg(const hs_proposal_message& msg); //send vote msg
+      void send_hs_vote_msg(const hs_vote_message& msg); //send proposal msg
+      void send_hs_new_view_msg(const hs_new_view_message& msg); //send new view msg
+      void send_hs_new_block_msg(const hs_new_block_message& msg); //send new block msg
+
+      void update(const hs_proposal_message& proposal); //update internal state
+      void commit(const hs_proposal_message& proposal); //commit proposal (finality)
+
+      void gc_proposals(uint64_t cutoff); //garbage collection of old proposals
+
+#warning remove. bls12-381 key used for testing purposes
+      //todo : remove. bls12-381 key used for testing purposes
+      std::vector<uint8_t> _seed =
+         {  0,  50, 6,  244, 24,  199, 1,  25,  52,  88,  192,
+            19, 18, 12, 89,  6,   220, 18, 102, 58,  209, 82,
+            12, 62, 89, 110, 182, 9,   44, 20,  254, 22 };
+
+      fc::crypto::blslib::bls_private_key _private_key = fc::crypto::blslib::bls_private_key(_seed);
+
+      enum msg_type {
+         new_view = 1,
+         new_block = 2,
+         qc = 3,
+         vote = 4
+      };
+
+      bool _chained_mode = false;
+      block_id_type _block_exec = NULL_BLOCK_ID;
+      block_id_type _pending_proposal_block = NULL_BLOCK_ID;
+      fc::sha256 _b_leaf = NULL_PROPOSAL_ID;
+      fc::sha256 _b_lock = NULL_PROPOSAL_ID;
+      fc::sha256 _b_exec = NULL_PROPOSAL_ID;
+      fc::sha256 _b_finality_violation = NULL_PROPOSAL_ID;
+      eosio::chain::quorum_certificate _high_qc;
+      eosio::chain::quorum_certificate _current_qc;
+      uint32_t _v_height = 0;
+      eosio::chain::extended_schedule _schedule;
+      base_pacemaker* _pacemaker = nullptr;
+      std::set<name> _my_producers;
```

Comment:
Why is this a `std::set` and not a `std::vector`. As far as I can tell we never use any of the `std::set` capabilities.

Reply:
Agree; there's no reason to sort or uniquify producer account names.